be/src/exec/vectorized/hdfs_scanner.cpp

```diff
@@ -179,8 +181,19 @@ uint64_t HdfsScanner::exit_pending_queue() {
 Status HdfsScanner::open_random_access_file() {
     CHECK(_file == nullptr) << "File has already been opened";
     ASSIGN_OR_RETURN(_raw_file, _scanner_params.fs->new_random_access_file(_scanner_params.path))
-    _file = std::make_unique<RandomAccessFile>(
-            std::make_shared<CountedSeekableInputStream>(_raw_file->stream(), &_stats), _raw_file->filename());
+    auto stream = std::make_shared<CountedSeekableInputStream>(_raw_file->stream(), &_stats);
+    if (_compression_type == CompressionTypePB::NO_COMPRESSION) {
+        _file = std::make_unique<RandomAccessFile>(stream, _raw_file->filename());
+    } else {
+        using DecompressorPtr = std::shared_ptr<StreamCompression>;
+        std::unique_ptr<StreamCompression> dec;
+        RETURN_IF_ERROR(StreamCompression::create_decompressor(_compression_type, &dec));
+        auto compressed_input_stream =
+                std::make_shared<io::CompressedInputStream>(stream, DecompressorPtr(dec.release()));
+        auto compressed_seekable_input_stream =
+                std::make_shared<io::CompressedSeekableInputStream>(compressed_input_stream);
+        _file = std::make_unique<RandomAccessFile>(compressed_seekable_input_stream, _raw_file->filename());
```

Comment:
So, if the derived classes of `HdfsScanner` support read compressed file it's sure that there will be no random accesses on this file, right? If this true, why do we have to create a `RandomAccessFile` but not `SequentialFile` for these classes?

Reply:
Nono, HdfsScanner indeed is supported to read random access file. It can read csv file from some offset and find synchronization point like line delimeter, and parse following text into records.

Reply:
Let me get it straight, hdfs text scanner can:
1. read uncompressed in random way by searching synchronization point like line delimeter.
2. read compressed in sequantial way, but have to read from begin to end.

So what's the underlying file should be? RandomAccessFile or SequentialFile?

Reply:
@sduzh What do you think?

Reply:
> Let me get it straight, hdfs text scanner can:
> 
> 1. read uncompressed in random way by searching synchronization point like line delimeter.
> 2. read compressed in sequantial way, but have to read from begin to end.
> 
> So what's the underlying file should be? RandomAccessFile or SequentialFile?

If so, can we create different files based on different situations?

Reply:
> > Let me get it straight, hdfs text scanner can:
> > 
> > 1. read uncompressed in random way by searching synchronization point like line delimeter.
> > 2. read compressed in sequantial way, but have to read from begin to end.
> > 
> > So what's the underlying file should be? RandomAccessFile or SequentialFile?
> 
> If so, can we create different files based on different situations?

Or, could you just implement a real `SeekableCompressedInputStream`?

Reply:
> > Let me get it straight, hdfs text scanner can:
> > 
> > 1. read uncompressed in random way by searching synchronization point like line delimeter.
> > 2. read compressed in sequantial way, but have to read from begin to end.
> > 
> > So what's the underlying file should be? RandomAccessFile or SequentialFile?
> 
> If so, can we create different files based on different situations?

Oh, that's hard. Think about this case, there are two csvreader class: one is holding SequentialFile, one is hold RandomAccessFile. But except that difference, all the rest code are the same.

Reply:
> > > Let me get it straight, hdfs text scanner can:
> > > 
> > > 1. read uncompressed in random way by searching synchronization point like line delimeter.
> > > 2. read compressed in sequantial way, but have to read from begin to end.
> > > 
> > > So what's the underlying file should be? RandomAccessFile or SequentialFile?
> > 
> > 
> > If so, can we create different files based on different situations?
> 
> Or, could you just implement a real `SeekableCompressedInputStream`?

what's real `SeekableCompressedInputStream`  ? To implement all APIs like `seek` and `get_size`? 

First of all, implement that is hard. And second, even if I can implement that, I think the performance will be so bad that people will stay from that.

Reply:
> > > Let me get it straight, hdfs text scanner can:
> > > 
> > > 1. read uncompressed in random way by searching synchronization point like line delimeter.
> > > 2. read compressed in sequantial way, but have to read from begin to end.
> > > 
> > > So what's the underlying file should be? RandomAccessFile or SequentialFile?
> > 
> > 
> > If so, can we create different files based on different situations?
> 
> Oh, that's hard. Think about this case, there are two csvreader class: one is holding SequentialFile, one is hold RandomAccessFile. But except that difference, all the rest code are the same.

If so, there should be something wrong in our API design. Anyway, since this PR has been merged I think it's ok to close this discussion now.

As a friendly reminder, please don't choose me as a reviewer if you decide to merge a PR without resolving my comments.

Reply:
> > > > Let me get it straight, hdfs text scanner can:
> > > > 
> > > > 1. read uncompressed in random way by searching synchronization point like line delimeter.
> > > > 2. read compressed in sequantial way, but have to read from begin to end.
> > > > 
> > > > So what's the underlying file should be? RandomAccessFile or SequentialFile?
> > > 
> > > 
> > > If so, can we create different files based on different situations?
> > 
> > 
> > Oh, that's hard. Think about this case, there are two csvreader class: one is holding SequentialFile, one is hold RandomAccessFile. But except that difference, all the rest code are the same.
> 
> If so, there should be something wrong in our API design. Anyway, since this PR has been merged I think it's ok to close this discussion now.
> 
> As a friendly reminder, please don't choose me as a reviewer if you decide to merge a PR without resolving my comments.

Sorry, this PR is in hurry and expected to be merged into branch-2.4 ASAP, so I don't resolve your comments. I did not intend to ignore your comments.  I apologize for that and I'lll follow your advice in future work.